MobiFlight/Joysticks/VKB/VKBDevice.cs

```diff
@@ -0,0 +1,192 @@
+                }
+            }
+            foreach (var encdec in EncoderDecList)
+            {
+                if (EncoderIncList.ContainsKey(encdec.Key))
+                {
+                    Encoders.Add(encdec.Key, new VKBEncoder(EncoderIncList[encdec.Key], encdec.Value)); // If both increment and decrement were in the definition, we create an encoder object from the definition file
+                    Buttons.Add(Encoders[encdec.Key].DeviceDec); // And register their virtual buttons in the joystick's button list.
+                    Buttons.Add(Encoders[encdec.Key].DeviceInc);
+                }
+            }
+        }
+
+        private static byte GetEncoderIndex(JoystickInput input)
+        {
+            return (byte)((input.Id - 1000) / 10);
+        }
+        private static VKBEncoder.EncoderAction GetEncoderAction(JoystickInput input)
+        {
+            return (VKBEncoder.EncoderAction)(input.Id % 10);
+        }
+        protected override void EnumerateOutputDevices()
+        {
+            Definition?.Outputs?.ForEach(output => Lights.AddChannel(output));
+            base.EnumerateOutputDevices();
+        }
+
+        public override void SetOutputDeviceState(string name, byte state)
+        {
+            Lights.UpdateState(name, state);
+        }
+
+
+        public override void UpdateOutputDeviceStates()
+        {
+            var data = Lights.CreateMessage();
+            if (data[7] == 0) return; // Only send message if there are non-zero LEDs to be updated
+            try
+            {
+                SendData(data);
+            }
+            catch (System.IO.IOException)
+            {
+                base.OnDeviceRemoved();
+            }
+        }
+        private void OnHidReportReceived(object sender, System.EventArgs e)
+        {
+            var inputReceiver = sender as HidDeviceInputReceiver;
+            while (inputReceiver.TryRead(InputReportBuffer, 0, out _))
+            {
+                byte ReportId = InputReportBuffer[0];
+                if (ReportId != 0x08) // 0x08 = Monitoring channel / virtual bus
+                    continue;
+                byte MessageType = InputReportBuffer[1];
+                if (MessageType != 0x13) // 0x13 = Encoder status
+                    continue;
+                ParseEncoderReport(InputReportBuffer);
+            }
+        }
+        private void ParseEncoderReport(byte[] Report)
+        {
+            byte sequenceNo = Report[2];
+            if (((lastSeqNo + 1) & 0xFF) != sequenceNo)
+            {
+                Log.Instance.log("Some VKB encoder messages may have been missed", LogSeverity.Debug);
+                // Can easily happen if many updates are sent in quick succession (like when an encoder is spun fast)
+                // Not a problem since API reports absolute state, so the updates are just received on the next received message.
```

Comment:
This comment should go above the if statement and explain that the if statement is checking to see if the sequency number increased by more than one from the last received message.